elastic/elastic-agent
# Add fleet-server secret files support

## What does this PR do?

Allow the user to specify the fleet-server's service_token through a file specified with service_token_file. And the ssl key passphrase through a file as well with an updated version of the tlscommon lib provided by elastic-agent-libs.

## Why is it important?

We do not want to expose fleet-server secrets like the key-passphrase or service_token through the command line or environment variables

## Checklist

- [x] I have made corresponding changes to the documentation [draft](https://github.com/elastic/observability-docs/pull/2839)
- ~~I have made corresponding change to the default configuration files~~
- ~~I have added an integration test or an E2E test~~

## How to test this PR locally

Run elasticsearch + kibana.

Add a fleet-server entry to Kibana, record the service token.

Create a fleet-server CA and certs with 
```
elasticsearch-certutil ca --pem --out stack.zip
unzip stack.zip
elasticsearch-certutil cert --name fleet-server --ca-cert ca/ca.crt --ca-key ca/ca.key --pem --ip $MY_IP --out cert.zip --pass abcd1234
```
Copy the CA cert and certs generated somewhere else (like `/tmp`)

Deploy secret files:
```
echo -n $SERVICE_TOKEN > /tmp/service-token
echo -n abcd1234 > /tmp/passphrase
```
**Note:** It's critical to ensure that no newline characters are appended in the secret files. 

Run the agent with:
```
sudo ./elastic-agent install \
  --url="https://$MY_IP:8220" \
  --certificate-authorities=/tmp/ca/ca.crt \
  --fleet-server-es="https://$ELASTICSEARCH_IP:9200" \
  --fleet-server-service-token-path=/tmp/service-token \
  --fleet-server-policy=fleet-server-policy \
  --fleet-server-es-ca-trusted-fingerprint=$ELASTICSEARCH_CA_FINGERPRINT \
  --fleet-server-cert=/tmp/fleet-server/fleet-server.crt \
  --fleet-server-cert-key=/tmp/fleet-server/fleet-server.key \
  --fleet-server-cert-key-passphrase=/tmp/passphrase
```

The agent/fleet-server instance should enrol and appear in the agents list of the UI.

## Related issues

- Requires https://github.com/elastic/fleet-server/pull/2459
